pkg/operator/ceph/disruption/clusterdisruption/osd.go

```diff
@@ -42,25 +40,83 @@ import (
 
 const (
 	// PDBAppName is that app label value for pdbs targeting osds
-	PDBAppName         = "rook-ceph-osd-pdb"
-	disabledPDBKey     = "disabled-pdb"
-	disabledPDBTimeKey = "pdb-disabled-at"
+	PDBAppName            = "rook-ceph-osd-pdb"
+	drainingFailureDomain = "draining-failureDomain"
+	drainingFDTimeKey     = "failureDomain-draining-since"
 	// DefaultMaintenanceTimeout is the period for which a drained failure domain will remain in noout
 	DefaultMaintenanceTimeout = 30 * time.Minute
 	nooutFlag                 = "noout"
 )
 
-func (r *ReconcileClusterDisruption) createPDBForOSD(deployment appsv1.Deployment) error {
+func (r *ReconcileClusterDisruption) createPDB(pdb *policyv1beta1.PodDisruptionBudget) error {
+	err := r.client.Create(context.TODO(), pdb)
+	if err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrapf(err, "failed to create pdb %q", pdb.Name)
+	}
+	return nil
+}
+
+func (r *ReconcileClusterDisruption) deletePDB(pdb *policyv1beta1.PodDisruptionBudget) error {
+	err := r.client.Delete(context.TODO(), pdb)
+	if err != nil && !kerrors.IsNotFound(err) {
+		return errors.Wrapf(err, "failed to create pdb %q", pdb.Name)
+	}
+	return nil
+}
+
+// createPDBforOSD creates a single PDB for all OSDs with maxUnavailable=1
+// This allows all OSDs in a single failure domain to go down.
+func (r *ReconcileClusterDisruption) createPDBforOSD(namespace string) error {
+	cephCluster, ok := r.clusterMap.GetCluster(namespace)
+	if !ok {
+		return errors.Errorf("failed to find the namespace %q in the clustermap", namespace)
+	}
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      PDBAppName,
+			Namespace: namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: cephCluster.APIVersion,
+					Kind:       cephCluster.Kind,
+					Name:       cephCluster.ObjectMeta.GetName(),
+					UID:        cephCluster.GetUID(),
+				},
+			},
+		},
+		Spec: policyv1beta1.PodDisruptionBudgetSpec{
+			MaxUnavailable: &intstr.IntOrString{IntVal: 1},
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{k8sutil.AppAttr: osd.AppName},
+			},
+		},
+	}
+	return r.createPDB(pdb)
+}
+
+func (r *ReconcileClusterDisruption) deletePDBforOSD(namespace string) error {
+	pdb := &policyv1beta1.PodDisruptionBudget{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      PDBAppName,
+			Namespace: namespace,
+		},
+	}
+	return r.deletePDB(pdb)
+}
+
+// createBlockingPDBForOSD creates individual blocking PDBs (maxUnavailable=0) for all the OSDs in
+// failure domains that are not draining
+func (r *ReconcileClusterDisruption) createBlockingPDBForOSD(deployment appsv1.Deployment) error {
```

Comment:
A PDB will be created for every OSD not in the failure domain being drained? I was expecting we would only need to create a PDB at the failure domain level. For example, if Zone A is being drained, Zone B and C would each have a single PDB created for maxUnavailable=0.

Reply:
update the code to create blocking PDB per failure domain rather than per OSD